enterprise/cypher/compiled-expressions/src/main/java/org/neo4j/cypher/internal/runtime/compiled/expressions/CypherFunctions.java

```diff
@@ -0,0 +1,279 @@
+import org.neo4j.internal.kernel.api.Transaction;
+import org.neo4j.values.AnyValue;
+import org.neo4j.values.SequenceValue;
+import org.neo4j.values.ValueMapper;
+import org.neo4j.values.storable.BooleanValue;
+import org.neo4j.values.storable.DateTimeValue;
+import org.neo4j.values.storable.DateValue;
+import org.neo4j.values.storable.DoubleValue;
+import org.neo4j.values.storable.DurationValue;
+import org.neo4j.values.storable.IntegralValue;
+import org.neo4j.values.storable.LocalDateTimeValue;
+import org.neo4j.values.storable.LocalTimeValue;
+import org.neo4j.values.storable.LongValue;
+import org.neo4j.values.storable.NumberValue;
+import org.neo4j.values.storable.PointValue;
+import org.neo4j.values.storable.TemporalValue;
+import org.neo4j.values.storable.TextValue;
+import org.neo4j.values.storable.TimeValue;
+import org.neo4j.values.storable.Value;
+import org.neo4j.values.storable.Values;
+import org.neo4j.values.virtual.ListValue;
+import org.neo4j.values.virtual.MapValue;
+import org.neo4j.values.virtual.PathValue;
+import org.neo4j.values.virtual.VirtualNodeValue;
+import org.neo4j.values.virtual.VirtualRelationshipValue;
+import org.neo4j.values.virtual.VirtualValues;
+
+import static org.neo4j.values.storable.Values.NO_VALUE;
+import static org.neo4j.values.storable.Values.doubleValue;
+import static org.neo4j.values.storable.Values.longValue;
+import static org.neo4j.values.storable.Values.stringValue;
+
+/**
+ * This class contains static helper methods for the set of Cypher functions
+ */
+@SuppressWarnings( "unused" )
+public final class CypherFunctions
+{
+    private CypherFunctions()
+    {
+        throw new UnsupportedOperationException( "Do not instantiate" );
+    }
+
+    public static Value sin( AnyValue in )
+    {
+        if ( in == NO_VALUE )
+        {
+            return NO_VALUE;
+        }
+        else if ( in instanceof NumberValue )
+        {
+            return doubleValue( Math.sin( ((NumberValue) in).doubleValue() ) );
+        }
+        else
+        {
+            throw new CypherTypeException( "sin requires numbers", null );
+        }
+    }
+
+    public static Value asin( AnyValue in )
+    {
+        if ( in == NO_VALUE )
+        {
+            return NO_VALUE;
+        }
+        else if ( in instanceof NumberValue )
+        {
+            return doubleValue( Math.asin( ((NumberValue) in).doubleValue() ) );
+        }
+        else
+        {
+            throw new CypherTypeException( "asin requires numbers", null );
+        }
+    }
+
+    public static Value haversin( AnyValue in )
+    {
+        if ( in == NO_VALUE )
+        {
+            return NO_VALUE;
+        }
+        else if ( in instanceof NumberValue )
+        {
+            return doubleValue( (1.0 - Math.cos( ((NumberValue) in).doubleValue() )) / 2 );
+        }
+        else
+        {
+            throw new CypherTypeException( "asin requires numbers", null );
```

Comment:
...although they might be disappointed in the error message when they mess up their types.